packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -288,157 +308,126 @@ export class OdspDocumentService implements IDocumentService {
-            const startAfd = performance.now();
-
-            return OdspDocumentDeltaConnection.create(
-                tenantId,
-                documentId,
-                token,
-                io,
-                client,
-                url2!,
-                20000,
-                this.logger,
-            ).then((connection) => {
-                logger.sendTelemetryEvent({
-                    eventName: "UsedAfdUrl",
-                    fromCache: true,
+            try {
+                const connection = await OdspDocumentDeltaConnection.create(
+                    tenantId,
+                    documentId,
+                    token,
+                    io,
+                    client,
+                    nonAfdUrl,
+                    20000,
+                    this.logger,
+                );
+                this.logger.sendTelemetryEvent({
+                    eventName: "NonAfdConnectionSuccess",
                 });
-
                 return connection;
-            }).catch(async (connectionError) => {
-                const endAfd = performance.now();
-                localStorage.removeItem(lastAfdConnectionTimeMsKey);
-                // Retry on non-AFD URL
-                if (this.canRetryOnError(connectionError)) {
-                    // eslint-disable-next-line max-len
-                    debug(`Socket connection error on AFD URL (cached). Error was [${connectionError}]. Retry on non-AFD URL: ${url}`);
-
-                    return OdspDocumentDeltaConnection.create(
-                        tenantId,
-                        documentId,
-                        token,
-                        io,
-                        client,
-                        url,
-                        20000,
-                        this.logger,
-                    ).then((connection) => {
-                        logger.sendPerformanceEvent({
-                            eventName: "UsedNonAfdUrlFallback",
-                            duration: endAfd - startAfd,
-                        }, connectionError);
-
-                        return connection;
-                    }).catch((retryError) => {
-                        logger.sendPerformanceEvent({
-                            eventName: "FailedNonAfdUrlFallback",
-                            duration: endAfd - startAfd,
-                        }, retryError);
-                        throw retryError;
-                    });
-                } else {
-                    logger.sendPerformanceEvent({
-                        eventName: "FailedAfdUrl-NoNonAfdFallback",
-                    }, connectionError);
-                }
+            } catch (connectionError) {
+                // Log before throwing
+                const canRetry = canRetryOnError(connectionError);
+                this.logger.sendTelemetryEvent(
+                    {
+                        eventName: "NonAfdConnectionFail",
```

Comment:
Since you are touching this code :)
I'd prefer different approach, The one where connectToDeltaStreamWithRetry() collects write data and stucks it on error object and throws it, without logging anything. And it's up to DeltaManager to reason about log or not to log it.
The reason I like it better is that we have central place to optimize amount of telemetry. I.e. today, for example, it will log error only of first try in a loop of retries, and this is good for offline scenario, where we keep retrying every 8 seconds, accumulating ton of telemetry that Aria flushes when client gets connected.

This does not solve telemetry needs for successful connection - I think it's fine for this code to issue an event in such cases with data on what was tried etc - number of these events will be much smaller than number of failures in disconnected state.

Reply:
I'd prefer not to attempt this in this PR.  Doing this in a manner that leaves DeltaManager agnostic to the particular driver likely requires establishing some new interfaces which is beyond the scope of this PR's goal, which is to get improved telemetry up and running to help with investigating OCE issue.  I'm also not entirely sure I like the idea of splitting the telemetry emission logic between layers of the stack like this.